### ISSUE TITLE
Add an option to prevent side panel from opening

### ISSUE THREAD

**Issue by rgarrigue** [microsoft/vscode#105270]
I've this kind of messages that keep popping up the side panel

![image](https://user-images.githubusercontent.com/1872031/91009682-60a43d00-e5e1-11ea-821d-08583053b0c0.png)

I guess it's a linter or whatever having small issues. There are multiples culprit, here it's "YAML Support", but I saw other ones. My issue is, I don't give a damn, and it's taking space, and popping up again and again if I close it. So far I could only stick it to the bottom and reduce it to the smallest size before closing... but now and then I like to open an integrated terminal, which lives in the same panel. Then the focus keep being stolen by those same messages.

So I'ld like a VS way to get rid of those. Asking plugin by plugin will never work, even if I knew where to report about those.

**Comment by sbatten** [comment]
request makes sense but really extensions should not aggressively show the channel. a setting would help the user now but any extensions behaving responsibly may not be able to reliably convey info.

**Comment by shankara-n** [comment]
Issue seems to be closed now but I can't find any setting to disable the output window from showing up

**Comment by JRodez** [comment]
"out-of-scope" they said...
So I'm going CLion if this kind of annoyance is OK for Microsoft.

**Comment by gjsjohnmurray** [comment]
@sandy081 please reopen this.

**Comment by jstm88** [comment]
> @sandy081 please reopen this.

I opened https://github.com/microsoft/vscode/issues/203478 with an extremely thorough explanation on why the issue is still relevant. @sandy081 closed it too, and appears to also be the one closing many of the others.

@sandy081 Do you have some sort of a vendetta against this specific feature? You've been closing these as duplicates even though it's *very* clear that the issue is not resolved. Users are trying to make their voice heard and you're almost single-handedly shutting down the discussions.

**Comment by jstm88** [comment]
The first time this issue was reported (that I can find) was https://github.com/microsoft/vscode/issues/33262 from August 2017. **Over 6 years ago.**

The fact that one was (incorrectly) closed has been used as a justification to close every other ticket addressing the problem, even though it has still not been resolved.

**Comment by jstm88** [comment]
So I guess that's it then?

Multiple people point out a major issue, write up detailed reports, and suggest good solutions, and one person can just come along and just shut down the report because of a failure to understand the issue?

And then that person just ignores all requests to reopen the ticket for multiple days, despite closing it instantly and without any sort of review?

You know, I really thought VS Code was better than that. But now the Microsoft-ness is coming through loud and clear.

**Comment by gjsjohnmurray** [comment]
Is #199946 going to mitigate this by allowing Output to be detached into its own window?

**Comment by jstm88** [comment]
@gjsjohnmurray No. That just shifts the problem elsewhere, and making an entire **window** pop up constantly would be even worse.

The issue is plugins being able to force the output panel to open. Many of them abuse (or misuse) the API to dump output constantly that is not useful. It's been said that the fault is with those plugins (and that's technically true) but in that way it's just like pop-up windows on the internet. We added pop-up blockers because they were so annoying. VS Code effectively needs the same thing for its own "pop-up output window" nuisance.

The argument has been made that "the output is important so we must allow it" - to which I say **who are you to decide what the user wants**? That's an insult to the users, who generally **do**, in fact, know better than Microsoft how they want their editor configured.

**Comment by gjsjohnmurray** [comment]
> allowing Output to be detached into its own window?

Actually I misinterpreted what the referenced change does. It merely opens a static copy of the output channel's contents. The ability to detach the entire Output view into its own window isn't with us yet.

**Comment by jstm88** [comment]
Still calling on @sandy081 (who closed this and then just completely abandoned us) to reopen one of these tickets.

**Comment by gjsjohnmurray** [comment]
> It is **not possible** to fix it any other way.

@jstm88 which of your installed extension is/are causing this? Have you raised issues with the extension author(s)? Or considered not using the extension(s)?

Somewhat ironic that your frequent comments on multiple issues are starting to feel to me like the kind of unwanted "spam" you're apparently getting from extensions you choose to install.

**Comment by jstm88** [comment]
@gjsjohnmurray There are multiple extensions that do this. One of the most common is *Microsoft's own* C extension.

The argument that we shouldn't do anything is flawed, and it's frustrating that a very small number of developers are using it as a weapon against users to shut down discussion.

I completely understand that, yes, this is technically the extensions' fault. However, with thousands of extensions and who knows how many of them misusing the API, it is absolutely hopeless to get everyone to fix every extension.

That's why we absolutely need an option to prevent focus-stealing on the Output window. There is no other way to fix this problem.

**Comment by jstm88** [comment]
@gjsjohnmurray Also worth noting that the only reason there are so many duplicates is because multiple issues were improperly closed, and people had to keep opening new tickets and wasting time trying to convince the mods to reopen them.

Given how many "duplicates" have been opened, it's pretty obvious that the users consider this to be a real problem despite what Microsoft might want to pretend.

**Comment by jstm88** [comment]
Specifically on the merits of blocking focus-stealing:

* It will only apply to users who enable the option, so there is no downside to having the option available.
* Badges will still indicate changes, as they do for other tabs. So there won't be a problem with errors being "hidden" as some have suggested. And even then, Microsoft has no right to tell me what errors I want to see or hide.
* Output will still be available and the panel will still behave exactly as it did before, it simply will not force itself to the front against the user's wishes.
* It does not prevent extensions from fixing their own problems. These solutions can be done in parallel. In fact, badging the output window (and, perhaps as a compromise, displaying a corner notification when the panel attempted to focus) results in a much cleaner experience *without* reducing user awareness.
* I reiterate, Microsoft has no right to tell me how to lay out my editor and what panels I need to see at any given moment. That's for me to decide and for VS Code to respect.

**Comment by gjsjohnmurray** [comment]
@jstm I think it is worth distinguishing "focus-stealing" from "showing the Output view". I think the latter can already be **mitigated** (word chosen carefully, as I support the broad direction of your proposal but want to be realistic about the development resources available in the core team, and the competing priorities).

An extension unilaterally showing the Output view is disruptive when that view is a tab on the Panel, because it steals visibility from whichever other tab the user had selected, and was perhaps actively working on.

By relocating the Output view this will no longer happen.

> One of the most common (and the worst offender on my setup) is Microsoft's own C extension.

Is there an linkable open issue on that extension's repo?

**Comment by gjsjohnmurray** [comment]
At the [API level](https://code.visualstudio.com/api/references/vscode-api#OutputChannel) the preserveFocus parameter of the show method can be used by an extension.

My reading of https://github.com/microsoft/vscode-cpptools/issues/7234 is that the developers were open to adding a setting to cover the case raised by the OP, but other changes were subsequently thought to have mitigated/resolved that case, and the OP didn't contest this view, so the issue was closed and is now locked.

I think if you opened a new issue in a non-confrontational way you might be able to make progress with regard to this particular extension.

More generally, it is useful to know that you have already moved Output view off Panel. I haven't yet investigated whether a show(true) call by an extension changes focus in this scenario. I wouldn't expect it to.

**Comment by gjsjohnmurray** [comment]
If it always calls `show()` rather than `show(true)` then it might be worth trying to get those calls changed.

**Comment by gjsjohnmurray** [comment]
Would it help any to move your Terminal into an editor tap, then move that into a floating window, and move Output back to Panel?

**Comment by jstm88** [comment]
@StringEpsilon That's a good point, perhaps the API needs to change `preserveFocus` to default to `true`. The no-arg version of something like should always default to the *least* intrusive mode.

That still leaves the possibility that extensions can abuse it and specify `false` - in that scenario we still would need a way to disable the output on a per-extension basis.

In fact, maybe if we think about the focus option as a **permission**, here's an even more comprehensive solution:

* The `show()` API defaults to `true` (i.e. the Output panel *does not* take focus by default)
* Extensions *may* specify `show(false)` in order to specifically override this and focus the panel
* Upon calling `show(false)` for the first time, VS Code requests **permission** from the user to allow the output panel to take focus **for that extension only**
* If the user denies permission, all calls from that extension act as if `preserveFocus` was set to `true`
* All extensions gain an option (true/false/unset=false): "Allow this extension to focus the Output panel"

It's a little more involved than a simple "disable" option, but I see some benefits:

* It defaults extensions to *not* being able to focus, which is actually even better
* It naturally exposes the option at the moment a new extension attempts to display "focused" output
* It fits better with current types of permission systems so it may be easier for some people to understand

**Comment by gjsjohnmurray** [comment]
Pending an acceptable solution (which is beyond my powers) I have been suggesting mitigations.

If you are not already using the Secondary Side Bar, or if you are using only one tab there, perhaps consign Output to its minimum size there.

Or an alternative non-floating home for your Terminal could be in an Editor Group below your main editor area.

**Comment by jstm88** [comment]
At a technical level, a temporary toggle should be effectively a one-liner. Something like this pseudocode:

```
if not (config.getBoolForKey("window.blockOutputPanelFocus") {
    ... proceed to focus the panel ...
}
```

In fact, even if the more complete permission-based system is adopted in the future, we could still use this in the meantime. Just do it like this:

1. Add the option to the next release. Users can add the option manually to our configs and get the benefits immediately.
2. Change the API default to true as soon as possible. Document this change, explain in the documentation that extensions should be extremely careful to limit usage of the focusing mode, and also document that a future change will require permissions before using that mode.
3. Work on the permissions system at whatever pace is warranted.
4. Release the permissions system and simultaneously deprecate the "window.blockOutputPanelFocus" key, which will now be unnecessary with the permissions controls in place.

**Comment by jstm88** [comment]
@gjsjohnmurray Are you going to reopen the issue?

**Comment by gjsjohnmurray** [comment]
> @gjsjohnmurray Are you going to reopen the issue?

As a contributor, not a team member, I don't have that power.

**Comment by jstm88** [comment]
I guess then that would be @sbatten or @sandy081 then?

Do we have to open a new ticket, or are we just being ghosted now?

**Comment by jstm88** [comment]
The January 2024 update came in and contains "Fine-grained control for disabling notifications per extension". It only makes logical sense, then, that we should also have the same control for Output from extensions. Addressing notifications is fine, but notifications are transient and can be ignored. The Output panel is worse because it disrupts the user by breaking the layout that has been specified.

Once again, @sbatten / @sandy081 someone needs to re-open this issue. It's been over 2 weeks without a single peep.

**Comment by gjsjohnmurray** [comment]
I have submitted PR #205225 that I hope will be considered by the team.

**Comment by jstm88** [comment]
@gjsjohnmurray That looks great!

From what I can see I think that covers everything, and it looks like the notification will be actionable as well. I can't imagine _anyone_ objecting to this.

I'm still miffed that the VS Code team refuses to even acknowledge the discussion. Being closed means these tickets were effectively hidden (since nobody really browses closed tickets unless they're looking for something specific). We're lucky @gjsjohnmurray was here with the knowledge needed. 👍 

So.. since there's an actual MR now, is *that* sufficient to reopen the ticket? Or do flying pigs need to be implemented first? 🤣

**Comment by jstm88** [comment]
@sbatten @sandy081 

Any updates? I know you're both on the PR but I'm not familiar with how long they usually take to get attention by a reviewer.

Also, while we're waiting on the PR to get reviewed, we still should re-open this issue. Then once the fix goes through the process it can be closed as complete, instead of permanently leaving it in this state. I know it's been left closed throughout this whole process but it really should be open so people can actually see that the issue is being addressed and aren't tempted to submit yet another duplicate.

**Comment by Gbox4** [comment]
This feature would be fantastic if added. Just adding my temporary fix was to drag the Output panel over to the secondary side bar on the right and making that as small as possible. Still annoying that it steals focus though, this silly annoyance should be easily fixable.

**Comment by jstm88** [comment]
@Gbox4 https://github.com/microsoft/vscode/pull/205225 by (the amazing!) @gjsjohnmurray fixes this, but Microsoft appears to be actively burying it.

If you look back through the years-long history of issues mentioning this problem, there are a handful of Microsoft employees doing their best to bury any reference to the issue and trying to tell everyone that it's not something that should be fixed The same two who have been hiding, closing, and ignoring all of the bug reports on this also appear to be doing the same with the PR now. I don't know why there's such a vendetta against a simple fix, but there you go.

Honestly, this episode (and **particularly** the _insulting_ disregard for @gjsjohnmurray 's work) has shown me that VS Code is too far gone.

For a while Microsoft looked to be going in a good direction (Azure had good Linux support, and VS Code was a good editor with a strong open source community). Unfortunately the direction VS Code has been going is way too close to the *other* side of Microsoft. Just look at what Microsoft is *currently* doing with Windows 11 for definitive proof of who Microsoft *really* is as a company. If Microsoft is still willing to do that, then it's hard even to rely on any "open source" projects by that same company. The behavior of Microsoft where they can get away with it makes any related project suspect as well.

Honestly, though, it's all down to @sandy081 and @sbatten failing to address the PR that has enlightened me to the state of VS Code. If this issue had just been open to discussion I really wouldn't have thought much of it. But by actively fighting the users (and even gaslighting them by trying to deny the issue exists) VS Code (as a project) has revealed its true colors.

Given all of that, I've decided it's too risky to continue to rely on VS Code. I've always had my eye on learning Vim motions properly (and I had planned on using them inside VS Code) but after seeing what's happened here I've decided to start the process of moving to Neovim instead. It's going to take some effort but in the long run I don't want to be relying on something that treats users like this. And as a bonus I won't have to deal with fixing things every month when the new update inevitably breaks something.

@gjsjohnmurray We appreciate the contribution you made here... even if Microsoft doesn't.

**Comment by starball5** [comment]
Related on Stack Overflow: [How to block a VS Code extension from opening output window on every parse error](https://stackoverflow.com/q/78496622/11107541)

**Comment by Refaat0** [comment]
pls fix this microsoft. or else im telling billy gates on you 😡😡😡😡

**Comment by benibenj** [comment]
I would like to see this problem resolved, whether by us directly addressing it or by encouraging the extension(s) responsible to implement a solution.

First, I want to fully understand the core issue. Specifically, I am trying to determine if this problem is confined to the Output view or if it affects revealing views in general. Could you please provide more details on which extensions you have encountered this problem with? I am aware of this being an issue with the C++ extension (as noted in [microsoft/vscode-cpptools#7234](https://github.com/microsoft/vscode-cpptools/issues/7234)). Are there other extensions experiencing the same issue?

Regarding potential solutions, the PR by @gjsjohnmurray ([#205225](https://github.com/microsoft/vscode/pull/205225)) seems promising. However, there might be alternative approaches, possibly akin to the notification settings. Let's explore all viable options to find the best resolution.

**Comment by VSCodeTriageBot** [comment]
:slightly_smiling_face: This feature request received a sufficient number of community upvotes and we moved it to our backlog. To learn more about how we handle feature requests, please see our [documentation](https://aka.ms/vscode-issue-lifecycle).

Happy Coding!

**Comment by benibenj** [comment]
Thanks for all your input! I think combining this with the `Do not Disturb` mode would make sense and would not require adding a new setting. It would also allow disabling the behaviour just for a specific set of extensions.

**Comment by gjsjohnmurray** [comment]
@benibenj I can see some sense in combining this with Do Not Disturb, but I don't think everyone who's interested in this issue (Panel appearing suddenly because extension forces its output channel to display) will want to set Do Not Disturb (i.e. suppress all non-error notifications) for an extension as the only way of suppressing Panel-popup.

Related - the only doc references I've been able to find for Do Not Disturb are in release notes:

- https://code.visualstudio.com/updates/v1_69#_do-not-disturb-mode
- https://code.visualstudio.com/updates/v1_86#_allow-disabling-notifications-per-extension

Perhaps its also time this feature got into the regular doc.

**Comment by gjsjohnmurray** [comment]
@benibenj any more thoughts on this? I'm willing improve my PR if it's deemed to need changes, e.g. greater configurability.

**Comment by benibenj** [comment]
We are not 100% sure yet how this should work. We are currently thinking it could behave the same way notification does. This would mean:
- enable/disable per extension
- not having a setting for this but storing in storage
- have a UI action to see the toggles
- block the reveal on the API level

Currently we are thinking not to have a global toggle but just per extension (no action like do not disturb for notifications)

Let me know if you have any thoughts about this approach. I still need to discuss this with @bpasero. Let me get back to you next week. Please ping me if I forget.

**Comment by benibenj** [comment]
Ok, lets start off with a single setting for the ouput channel, similar to what is suggested in the PR [#205225](https://github.com/microsoft/vscode/pull/205225);

**Comment by gjsjohnmurray** [comment]
The new `workbench.view.showQuietly` setting is now available in today's Insiders.

**Comment by benibenj** [comment]
Verification steps: https://github.com/microsoft/vscode/pull/205225#issuecomment-2485356019

**Comment by alexr00** [comment]
Verified, but that status bar indicator is very subtle.

**Comment by jstm88** [comment]
> Verified, but that status bar indicator is very subtle.

When I see “but…” and “…very subtle” together, that tells me it’s done right. 😁

The editor’s job is to _get out of my way_ so I can code. The amount of bloat (especially with the AI fad) is insane, so everything we can do to hide/remove the cruft is a major win.
